### ISSUE TITLE
KeeAgent usability issues with window "Select a key" (Linux)

### ISSUE THREAD

**Issue by taurus-forever** [dlech/KeeAgent#157]
Hi,

I am experiencing some KeeAgent usability issues in my Linux environment and decided to share all of them hoping some will be addressed or workarounded.

So, environment:
Linux Mint 17.3
KeePass 2.33
KeeAgent 0.7.12  in Agent mode
Mono 3.2.8 (Debian 3.2.8+dfsg-4ubuntu1.1)

Everything started from the state when you have a lot of servers AND a lot of different keys.

Using the normal SSH configuration you can use the option IdentityFile for every host, like:

```
Host example.com
  IdentityFile ~/.ssh/id_rsa.example.com
```

Using KeeAgent as ssh Agent mode illuminate this option. Isn't it?

So, you have to load all the keys. If you have more then 5 keys server closes connection after the 5 tries, like described in https://github.com/dlech/KeeAgent/issues/139

At this stage you have several options:
- play the game with load/unload key manually
- set checkbox "Show key selection dialog when a client program requests a list of all keys"

The first option it too boring after the first several days of use.
The second option is looks OK, while it has several usability problems when new connection is established and "Select a key" is displayed:

1) pressing enter selects wrong key if key is selected with keyboard cursor.
If I select 3rd key (by pressing 2 times button down and then enter) the 2nd key is returned by sshagent. o_0 Can you reproduce it?
If Alt+S is pressed to confirm key selection by keyboard - all OK.
selection by mouse double click - works well.
Esc button for canceling works well.

2) "Select a key" window is always opened with column width minimized.
So, I have to expand column width size, then choose a key by proper column comment.
On the next use of "Select a key" window, then columns are minimized again. :-(

3) every time when SSH requests a key from KeeAgent, the "Select a key" window having "KeePass" window on background. As a result you have to switch back from window Keepass to SSH console after key selection. :-(

4) Order of columns in window "Select a key" is not really usable, normally you use a "Comment" or "Source" columns to identify which key to use. Unfortunately they are last one, so using some small window requires horizontal scrolling to see the most right columns. 
IMHO they should be reordered as type/size/fingerprint a barely usable on this stage.

5) window "Select a key" can be opened multiply times which confuses.
STR: ssh1 requests a key, poppup window "Select a key" is opened, ignore it, switch back to ssh1 and interrupt it by Ctrl+C  then request new ssh2 session -> second   poppup window "Select a key" is opened. You can repeat this game more and more. Minor-minor issue, just noticed.

Please consider to improve usability here.
Maybe I miss something and there is much better way to handle the case of huge keys list, please share if so. Thank you!

P.S. I cannot find a donate button on your site.
Your plugins work very well for me for a years,
so I would like to share some coffee/beer with you.

**Comment by taurus-forever** [comment]
You do not need to pass from SSH to KeeAgent any secret connection related data, it can be some marker from ssh config, like:

```
Host example.com
  SSHAgentIdentity blabla
```

5 is not a problem at all, it just confused me a bit for the first time.
You may catch it on KeeAgent level (force closing all windows if new request came, etc).
While you can create more problems here. Forget. JFYI, only.

Having a nice usability select window is something we can start.

---

> Sorry about that...PayPal.Me isn't available in your country.

Austria.

FAQ:
https://www.paypal.me/pages/faqs
> Is PayPal.Me available in my country?
> We’re adding new countries as fast as we can. Check back often to see if it’s available in your country.

:-( Ping me if you will be nearby.

**Comment by taurus-forever** [comment]
Also accordingly to http://www.unixwiz.net/techtips/ssh-agent-forwarding.html
SSH agent (KeeAgent) should receive "key challenge" from local ssh client.
Is it not "stable" to be saved somewhere in KeePass as a marker to choose a key?

**Comment by taurus-forever** [comment]
Oh and playing more here noticed new usability problem:

6) It is possible to select multiply keys in the form (using shift), which is great, I didn't aware of it!
While it is NOT impossible to select a random keys (using common Ctrl + mouse click).
It is possible to select only the keys nearby.

**Comment by dlech** [comment]
> I'd love to just be able to supply a -i filename option on the command line, but that doesn't work.

This should work. What have you tried that does not work?

**Comment by dlech** [comment]
> Also accordingly to http://www.unixwiz.net/techtips/ssh-agent-forwarding.html
> SSH agent (KeeAgent) should receive "key challenge" from local ssh client.
> Is it not "stable" to be saved somewhere in KeePass as a marker to choose a key?

What that page doesn't mention is that before "SSH agent (KeeAgent) should receive "key challenge" from local ssh client" if you do not use the `-i` option the ssh client program first requests a list of *all* ssh keys. The "key challenge" part just means that the ssh client program sends the fingerprint of the key that should be used to sign the request.

We already can figure out the fingerprint of every key in the KeePass database, so storing the fingerprint sent by the ssh client program would not do anything useful.

**Comment by taurus-forever** [comment]
>> I'd love to just be able to supply a -i filename option on the command line, but that doesn't work.
> This should work. What have you tried that does not work?
It works well if you have key as local file. It doesn't work if you store keys inside keepass.
You cannot point ssh to use some specific key from keepass/ssh-agent. Isn't it?

**Comment by dlech** [comment]
There is an option to save keys from KeePass to a file when the key is loaded.

![selection_009](https://cloud.githubusercontent.com/assets/963645/16322649/383d974e-396a-11e6-957e-a40bd5783064.png)

**Comment by taurus-forever** [comment]
Yes it can be the option, while it illuminates all the benefits storing keys in keepass.
(keys may stay in tmp location if keepass crash, etc).

**Comment by dlech** [comment]
> (keys may stay in tmp location if keepass crash, etc).

Yes, this is a problem, but if you key file has strong encryption, it is not so bad.

> Also in this case I do not need keeagent in agent mode anymore :-)

KeeAgent still takes care of entering the passphrase for you.

-----

But, yes, I understand your point this is not a super great option.

**Comment by taurus-forever** [comment]
OFFTOP!

re:
> Sorry about that...PayPal.Me isn't available in your country.

In light of:
>  As a first step, I have setup a PayPal payment button on the KeeAgent website...

Tested new button, works well! Waiting for KeeAgent 0.10.0 :-)

re2: ev3dev, I decided to decline lego ev3 for my son for the moment, it is too expensive toy which actually teaches mostly nothing out of the box. While hardware quality is really nice there.
Now playing with Elegoo Smart Robot Car, JFYI only.

**Comment by dlech** [comment]
Fix has been released in [KeeAgent 0.9.0](http://lechnology.com/2017/01/keeagent-beta-0-9-0-released/). Please test.

If there was something I missed, please open a new issue (one issue per suggestion).
